fdb-extensions/src/main/java/com/apple/foundationdb/async/RankedSet.java

```diff
@@ -246,6 +382,80 @@ public RankedSet(Subspace subspace, Executor executor) {
+                            final boolean duplicate = count > 1;
+                            final int nlevels = config.getNLevels();
+                            final List<CompletableFuture<Void>> futures = new ArrayList<>(nlevels);
+                            for (int li = 0; li < nlevels; ++li) {
+                                final int level = li;
+
+                                final CompletableFuture<Void> future;
+
+                                if (duplicate) {
+                                    // Always subtract one, never clearing a key. It is possible for this to leave
+                                    // a key with a count of zero, if duplicates were inserted with different hash functions.
+                                    Function<byte[], Void> decrement = k -> {
+                                        tr.mutate(MutationType.ADD, subspace.pack(Tuple.from(level, k)), encodeLong(-1));
+                                        return null;
+                                    };
+                                    if (level == 0) {
+                                        decrement.apply(key);
+                                        future = DONE;
+                                    } else {
+                                        future = getPreviousKey(tr, level, key, true).thenApply(decrement);
+                                    }
+                                } else {
+                                    // This could be optimized to check the hash for which levels should have this key.
+                                    // That would require that the hash function never changes, though.
+                                    // This allows for it to change, with the distribution perhaps getting a little uneven
+                                    // as a result. It even allows for the hash function to return a random number.
+                                    final byte[] k = subspace.pack(Tuple.from(level, key));
+                                    final CompletableFuture<byte[]> cf = tr.get(k);
+
+                                    if (level == 0) {
+                                        future = cf.thenApply(c -> {
+                                            if (c != null) {
+                                                tr.clear(k);
+                                            }
+                                            return null;
+                                        });
+                                    } else {
+                                        final CompletableFuture<byte[]> prevKeyF = getPreviousKey(tr, level, key, false);
+                                        future = CompletableFuture.allOf(cf, prevKeyF)
+                                                .thenApply(vignore -> {
+                                                    final byte[] c = cf.join();
+                                                    long countChange = -1;
```

Comment:
Not much thought went into mixing set and multiset on the same subspace. Other than not violating the consistency invariants. In particular, `remove` always removes (up to) one occurrence. Would it be better if it removed _all_ when not counting duplicates?

Reply:
I fixed the comments. It should not, in fact, be possible to decrement to zero any more.

Which means that if we did decide that `remove` should mean remove all, it would be consistent in clearing any splits induced by the given key.

Reply:
Maybe I'm missing the change, but it looks like one could still decrement a key to zero (in duplicate counting mode), in that it just decrements the value by 1.

Reply:
In `add`, if a key is already present (based on checking level 0 != 0), then the hash is ignored and each key that was incremented last time is incremented again. In other words, a duplicate value is counted n times at every level reliably.

In `remove`, if a key is not the last occurrence (based on checking level 0 > 1), then it is safe to simply decrement every level because this key must have contributed at least 2 to each of them. If a key is the last occurrence, then if the key decremented at any level is matches exactly (as opposed to being somewhere to the left), it is cleared. This is the only key that might otherwise have been decremented to zero, I believe.

Simultaneous `remove`s of the same key conflict on level 0.

Reply:
Oh, I think I see what I was missing: I had though that the `duplicate` boolean was determined by the configuration rather than by the count that actually gets read. I'll have to rethink what this is doing with that in mind.

Reply:
Yeah, this now makes sense to me, and I believe you're right--that essentially because all updates contend on their level 0 key, some of the nasty edge cases that can lead to concurrent decrements resulting in there being a count set to zero. I suppose we may need to re-architect if we decide that that's "too much contention" or something.

Reply:
As to the original question, I *think* that having remove always be "remove 1" is fine, and now here my reasoning is: (1) having it that way decreases the number of branches (on the config value) from the code and (2) changing the value of the config parameter is weird enough that we should generally avoid doing it (hence banning changing it at the meta-data level) and shouldn't add complexity to the code to support it.

But my feelings aren't particularly strong.